tests_tf/test_attacks.py

```diff
@@ -255,7 +255,7 @@ def test_generate_np_can_be_called_with_different_eps(self):
                                       clip_min=-5.0, clip_max=5.0)
 
       delta = np.max(np.abs(x_adv - x_val), axis=1)
-      self.assertClose(delta, eps)
+      self.assertTrue(np.all(delta<=(eps+1e-4)))
```

Comment:
This looks like it's actually changing the properties of the test.

Can you put a copy of the assertClose version in TestFastGradientMethod? The constraint should aways be tight for that attack.

What happened here is first we wrote the test for FastGradientMethod, then other people subclassed it for other attacks, then I renamed it to intermediate CommonAttackProperties class and pushed things that are specific to fast gradient method into a new TestFastGradientMethod subclass.

Reply:
It might be better to have one helper function that generates the adversarial examples and then have FGSM test that it's on the boundary and then PGD test that it's in the range [-eps,eps], not just exactly at the boundary.